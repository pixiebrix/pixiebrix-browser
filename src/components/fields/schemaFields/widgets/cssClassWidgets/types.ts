/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Expression } from "@/types/runtimeTypes";
import type React from "react";

export type Spacing = {
  side: string | null;
  size: number;
};

export type Value = string | Expression;

/**
 * An independent class name
 */
export type ClassFlag = {
  /**
   * The Bootstrap 4 class name
   */
  className: string;

  /**
   * Title node to render for the element (in a button/dropdown)
   */
  title: React.ReactNode;

  /**
   * True if the flag is exclusive for it's group (default=true)
   */
  exclusive?: boolean;

  /**
   * Other flags in the same group that the flag implies
   */
  implies?: string[];
};
