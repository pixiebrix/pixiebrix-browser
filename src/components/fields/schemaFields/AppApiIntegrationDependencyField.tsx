/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import type React from "react";
import { useField, useFormikContext } from "formik";
import { PIXIEBRIX_INTEGRATION_ID } from "@/services/constants";
import { produce } from "immer";
import { isEqual, set } from "lodash";
import {
  keyToFieldValue,
  type IntegrationsFormSlice,
} from "./integrations/integrationDependencyFieldUtils";
import {
  type Expression,
  type OutputKey,
  type ServiceVarRef,
} from "@/types/runtimeTypes";
import useAsyncEffect from "use-async-effect";

const PIXIEBRIX_OUTPUT_KEY = "pixiebrix" as OutputKey;

/**
 * Schema-based field for the PixieBrix API (@pixiebrix/api).
 *
 * - Does not render a DOM element
 * - Ensures a @pixiebrix/api dependency is included for the extension
 */
const AppApiIntegrationDependencyField: React.FunctionComponent<{
  name: string;
}> = ({ name }) => {
  const { values: root, setValues: setRootValues } =
    useFormikContext<IntegrationsFormSlice>();
  const [
    { value: dependencyOutputKey },
    ,
    { setValue: setDependencyOutputKey },
  ] = useField<Expression<ServiceVarRef>>(name);

  // This currently happens when a brick is copy-pasted into a separate extension
  // that does not yet have root.integrationDependencies configured, but already has the
  // integration dependency key set up in the (copied) BrickConfig.
  const isBadValue =
    dependencyOutputKey &&
    !root.integrationDependencies.some(({ outputKey }) =>
      isEqual(keyToFieldValue(outputKey), dependencyOutputKey)
    );

  useAsyncEffect(
    async () => {
      if (dependencyOutputKey == null) {
        const match = root.integrationDependencies.find(
          ({ integrationId }) => integrationId === PIXIEBRIX_INTEGRATION_ID
        );
        if (match?.outputKey) {
          // If the service is already being used, default to the currently configured auth
          console.debug(
            "PixieBrix dependency already exists for %s, using output key %s",
            match.integrationId,
            match.outputKey,
            { root, match }
          );
          await setDependencyOutputKey(keyToFieldValue(match.outputKey));
        } else {
          console.debug("Adding PixieBrix API dependency");
          await setRootValues(
            produce(root, (draft) => {
              draft.integrationDependencies.push({
                integrationId: PIXIEBRIX_INTEGRATION_ID,
                // XXX: in practice the pixiebrix outputKey won't be used by any other service. However, we might
                // consider using fresh identifier here to eliminate the possibility of colliding with a different
                // service that is somehow already using the @pixiebrix key
                outputKey: PIXIEBRIX_OUTPUT_KEY,
                // PixieBrix service does not use an ID -- the auth is automatically handled based on the logged-in user
                // configId: undefined,
              });

              set(draft, name, keyToFieldValue(PIXIEBRIX_OUTPUT_KEY));
            })
          );
        }
      } else if (isBadValue) {
        // Clearing this "bad value" value will enable the preceding if-branch to
        // execute again, and that will configure root.services properly
        await setDependencyOutputKey(null);
      }
    },
    // Run on mount, or if we detect a "bad value" (see comment above)
    [isBadValue]
  );

  return null;
};

export default AppApiIntegrationDependencyField;
