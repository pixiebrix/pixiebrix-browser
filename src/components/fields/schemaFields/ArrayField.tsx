/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Ok to use object here since we don't have any key-specific logic
// eslint-disable-next-line @typescript-eslint/ban-types
import React from "react";
import { FieldProps } from "@/components/fields/propTypes";
import { FieldArray, useField } from "formik";
import { Schema } from "@/core";
import { Button, Form } from "react-bootstrap";
import { fieldLabel } from "@/components/fields/fieldUtils";
import { getDefaultField } from "@/components/fields/blockOptions";
import { booleanPredicate, findOneOf, textPredicate } from "./schemaUtils";

function getDefaultArrayItem(schema: Schema): unknown {
  if (schema.default) {
    return schema.default;
  }

  if (textPredicate(schema)) {
    return "";
  }

  if (schema.type === "object") {
    return {};
  }

  if (findOneOf(schema, booleanPredicate)) {
    return false;
  }

  if (findOneOf(schema, textPredicate)) {
    return "";
  }

  return null;
}

const ArrayField: React.FunctionComponent<FieldProps<object[]>> = ({
  schema,
  label,
  ...props
}) => {
  const [field] = useField(props);

  if (Array.isArray(schema.items)) {
    throw new TypeError("Support for arrays of mixed types is not implemented");
  } else if (typeof schema.items === "boolean") {
    throw new TypeError("Schema required for items");
  }

  const schemaItems = schema.items as Schema;

  return (
    <Form.Group controlId={field.name}>
      <Form.Label>{label ?? fieldLabel(field.name)}</Form.Label>
      {schema.description && (
        <Form.Text className="text-muted">{schema.description}</Form.Text>
      )}
      <FieldArray name={field.name}>
        {({ remove, push }) => (
          <>
            <ul className="list-group">
              {(field.value ?? []).map((item: unknown, index: number) => {
                const Renderer = getDefaultField(schemaItems);
                return (
                  <li className="list-group-item" key={index}>
                    <Renderer
                      key={index}
                      name={`${field.name}.${index}`}
                      schema={schemaItems}
                      label={`${label ?? fieldLabel(field.name)} #${index + 1}`}
                    />
                    <Button
                      variant="danger"
                      size="sm"
                      onClick={() => {
                        remove(index);
                      }}
                    >
                      Remove Item
                    </Button>
                  </li>
                );
              })}
            </ul>
            <Button
              onClick={() => {
                push(getDefaultArrayItem(schemaItems));
              }}
            >
              Add Item
            </Button>
          </>
        )}
      </FieldArray>
    </Form.Group>
  );
};

export default ArrayField;
