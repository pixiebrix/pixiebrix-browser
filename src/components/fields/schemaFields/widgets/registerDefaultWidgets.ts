/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UrlMatchPatternWidget from "./UrlMatchPatternWidget";
import UrlPatternWidget from "@/pageEditor/components/UrlPatternWidget";
import ArrayWidget from "./ArrayWidget";
import BooleanWidget from "./BooleanWidget";
import IntegerWidget from "./IntegerWidget";
import NumberWidget from "./NumberWidget";
import ObjectWidget from "./ObjectWidget";
import OmitFieldWidget from "./OmitFieldWidget";
import SchemaSelectWidget from "./SchemaSelectWidget";
import TemplateToggleWidget from "./TemplateToggleWidget";
import TextWidget from "./TextWidget";
import UnsupportedWidget from "./UnsupportedWidget";
import widgetsRegistry from "./widgetsRegistry";
import WorkshopMessageWidget from "./WorkshopMessageWidget";
import SelectorMatchWidget from "@/pageEditor/components/SelectorMatchWidget";
import CssClassWidget from "./CssClassWidget";
import DatabaseWidget from "./DatabaseWidget";
import PasswordWidget from "./PasswordWidget";
import IconWidget from "@/components/fields/IconWidget";

const defaultWidgets = {
  ArrayWidget,
  BooleanWidget,
  IntegerWidget,
  NumberWidget,
  ObjectWidget,
  OmitFieldWidget,
  SchemaSelectWidget,
  TemplateToggleWidget,
  TextWidget,
  UnsupportedWidget,
  UrlMatchPatternWidget,
  UrlPatternWidget,
  SelectorMatchWidget,
  CssClassWidget,
  WorkshopMessageWidget,
  DatabaseWidget,
  IconWidget,
  PasswordWidget,
} as const;

function registerDefaultWidgets() {
  Object.assign(widgetsRegistry, defaultWidgets);
  for (const [name, widget] of Object.entries(defaultWidgets)) {
    if (!widget) {
      throw new Error(
        `Error registering default widgets. ${name} is undefined. Is there a circular dependency?`
      );
    }
  }
}

export default registerDefaultWidgets;
