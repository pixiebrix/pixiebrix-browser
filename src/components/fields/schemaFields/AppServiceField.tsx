/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useEffect } from "react";
import { SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { useField, useFormikContext } from "formik";
import { Expression, OutputKey, ServiceKeyVar } from "@/core";
import { PIXIEBRIX_SERVICE_ID } from "@/services/constants";
import {
  keyToFieldValue,
  ServiceSlice,
} from "@/components/fields/schemaFields/ServiceField";
import { produce } from "immer";
import { createTypePredicate } from "@/components/fields/fieldUtils";
import { SERVICE_BASE_SCHEMA } from "@/services/serviceUtils";
import { isEqual } from "lodash";

const PIXIEBRIX_OUTPUT_KEY = "pixiebrix" as OutputKey;

export const isAppServiceField = createTypePredicate(
  (schema) => schema.$ref === `${SERVICE_BASE_SCHEMA}${PIXIEBRIX_SERVICE_ID}`
);

/**
 * Schema-based field for the PixieBrix API (@pixiebrix/api).
 *
 * - Does not render a DOM element
 * - Ensures a @pixiebrix/api dependency is included for the extension
 */
const AppServiceField: React.FunctionComponent<SchemaFieldProps> = ({
  schema,
  uiSchema,
  ...props
}) => {
  const {
    values: root,
    setValues: setRootValues,
  } = useFormikContext<ServiceSlice>();
  const [{ value }, , helpers] = useField<Expression<ServiceKeyVar>>(props);

  // This currently happens when a brick is copy-pasted into a separate extension
  // that does not yet have root.services configured, but already has the service
  // key set up in the (copied) BlockConfig.
  const isBadValue =
    value &&
    !root.services.some((service) =>
      isEqual(keyToFieldValue(service.outputKey), value)
    );

  useEffect(
    () => {
      if (value == null) {
        const match = root.services.find(
          (service) => service.id === PIXIEBRIX_SERVICE_ID
        );
        if (match?.outputKey) {
          // If the service is already being used, default to the currently configured auth
          console.debug(
            "PixieBrix dependency already exists for %s, using output key %s",
            match.id,
            match.outputKey,
            { root, match }
          );
          helpers.setValue(keyToFieldValue(match.outputKey));
        } else {
          console.debug("Adding PixieBrix API dependency");
          // Try defaulting to the only option available. Use onChange instead of helpers.setValue b/c it automatically
          // updates the services part of the form state
          setRootValues(
            produce(root, (draft) => {
              draft.services.push({
                id: PIXIEBRIX_SERVICE_ID,
                // XXX: in practice it the pixiebrix outputKey won't be used by any other service. However, we might
                // consider using fresh identifier here to eliminate the possibility of colliding with a different
                // service that is somehow already using the @pixiebrix key
                outputKey: PIXIEBRIX_OUTPUT_KEY,
                // PixieBrix service does not use an ID -- the auth is automatically handled based on the logged in user
                config: null,
              });
            })
          );
          helpers.setValue(keyToFieldValue(PIXIEBRIX_OUTPUT_KEY));
        }
      } else if (isBadValue) {
        // Clearing this "bad value" value will enable the preceding if-branch to
        // execute again, and that will configure root.services properly
        helpers.setValue(null);
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps -- run on mount, or if we detect a "bad value" (see comment above)
    [isBadValue]
  );

  return null;
};

export default AppServiceField;
