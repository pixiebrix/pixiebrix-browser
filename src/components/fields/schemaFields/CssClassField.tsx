/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import CssClassWidget from "@/components/fields/schemaFields/widgets/cssClassWidgets/CssClassWidget";
import { getToggleOptions } from "@/components/fields/schemaFields/getToggleOptions";
import { type SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import defaultFieldFactory from "@/components/fields/schemaFields/defaultFieldFactory";

const RawCssClassField = defaultFieldFactory(CssClassWidget);

// Can't be constant because getToggleOptions needs the widgets registry to be initialized
export const getCssClassInputFieldOptions = () =>
  getToggleOptions({
    fieldSchema: {
      type: "string",
    },
    isRequired: false,
    customToggleModes: [],
    isArrayItem: false,
    isObjectProperty: false,
    allowExpressions: true,
  });

const CssClassField: React.VFC<SchemaFieldProps> = (props) => {
  const cssClassInputFieldOptions = useMemo(
    () => getCssClassInputFieldOptions(),
    []
  );

  const enrichedProps = {
    ...props,
    inputModeOptions: cssClassInputFieldOptions,
  } as SchemaFieldProps;

  return <RawCssClassField {...enrichedProps} />;
};

export default CssClassField;
