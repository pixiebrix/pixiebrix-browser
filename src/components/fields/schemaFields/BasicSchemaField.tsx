/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { type SchemaFieldComponent } from "@/components/fields/schemaFields/propTypes";
import { makeLabelForSchemaField } from "@/components/fields/schemaFields/schemaFieldUtils";
import SchemaFieldContext from "@/components/fields/schemaFields/SchemaFieldContext";
import { useField } from "formik";
import { isEmpty } from "lodash";
import FieldTemplate from "@/components/form/FieldTemplate";
import cx from "classnames";
import FieldRuntimeContext from "@/components/fields/schemaFields/FieldRuntimeContext";
import { getToggleOptions } from "./getToggleOptions";
import widgetsRegistry from "./widgets/widgetsRegistry";
import useToggleFormField from "@/hooks/useToggleFormField";
import { getFieldValidator } from "@/components/fields/fieldUtils";
import useFieldAnnotations from "@/components/form/useFieldAnnotations";
import { isExpression } from "@/utils/expressionUtils";
import useAsyncEffect from "use-async-effect";
import { type InputModeOption } from "@/components/fields/schemaFields/widgets/templateToggleWidgetTypes";

/*
 *  This is a hack to fix the issue where the formik state is not updated correctly when the form is first rendered.
 *  We use the renderRef to ensure that we only run this on the second render.
 *  TODO: We should be setting the initialValues in Redux before rendering the form.
 */
function useSetInitialValueForField({
  name,
  isRequired,
  inputModeOptions,
}: {
  name: string;
  isRequired: boolean;
  inputModeOptions: InputModeOption[];
}) {
  const renderRef = useRef(false);
  const [{ value }, , { setValue }] = useField(name);

  useEffect(() => {
    renderRef.current = true;
  }, []);

  useAsyncEffect(async () => {
    // Initialize any undefined required fields to prevent inferring an "omit" input
    if (
      value === undefined &&
      isRequired &&
      !isEmpty(inputModeOptions) &&
      renderRef.current
    ) {
      await setValue(inputModeOptions[0].interpretValue(value));
    }
    // We include setValue in the dependencies becuase sometimes the formik
    // helpers reference (setValue) changes, so we need to account for that in the dependencies
    // See: https://github.com/pixiebrix/pixiebrix-extension/issues/2269
  }, [setValue, renderRef.current]);
}

const BasicSchemaField: SchemaFieldComponent = ({
  omitIfEmpty = false,
  onBlur: onBlurProp,
  ...restProps
}) => {
  const {
    name,
    schema,
    validationSchema,
    isRequired,
    description,
    isObjectProperty = false,
    isArrayItem = false,
    hideLabel,
  } = restProps;
  const fieldLabel = makeLabelForSchemaField(restProps);
  const defaultDescription = useMemo(
    () => description ?? schema.description,
    [description, schema.description]
  );
  const [fieldDescription, setFieldDescription] =
    useState<React.ReactNode>(defaultDescription);

  const updateFieldDescription = useCallback(
    (newDescription: React.ReactNode | undefined) => {
      setFieldDescription(newDescription ?? defaultDescription);
    },
    [defaultDescription]
  );

  const { customToggleModes } = useContext(SchemaFieldContext);
  const { allowExpressions } = useContext(FieldRuntimeContext);

  const normalizedSchema = useMemo(() => {
    const isObjectType =
      schema.type === "object" || !Object.hasOwn(schema, "type");

    if (
      isObjectType &&
      schema.properties === undefined &&
      schema.additionalProperties === undefined &&
      schema.oneOf === undefined &&
      schema.anyOf === undefined &&
      schema.allOf === undefined
    ) {
      return {
        ...schema,
        additionalProperties: true,
      };
    }

    return schema;
  }, [schema]);

  const inputModeOptions = useMemo(
    () =>
      getToggleOptions({
        fieldSchema: normalizedSchema,
        isRequired,
        customToggleModes,
        isObjectProperty,
        isArrayItem,
        allowExpressions,
      }),
    [
      normalizedSchema,
      isRequired,
      customToggleModes,
      isObjectProperty,
      isArrayItem,
      allowExpressions,
    ]
  );

  const validate = getFieldValidator(validationSchema);

  const [{ value, onBlur: formikOnBlur }, { touched }] = useField({
    name,
    validate,
  });

  const annotations = useFieldAnnotations(name);

  useSetInitialValueForField({ name, isRequired, inputModeOptions });

  const { onOmitField } = useToggleFormField(name, normalizedSchema);

  if (isEmpty(inputModeOptions)) {
    return (
      <FieldTemplate
        name={name}
        label={fieldLabel}
        description={fieldDescription}
        annotations={annotations}
        touched={touched}
        as={widgetsRegistry.UnsupportedWidget}
      />
    );
  }

  const onBlur = (event: React.FocusEvent) => {
    formikOnBlur(event);

    if (
      omitIfEmpty &&
      (isEmpty(value) || (isExpression(value) && isEmpty(value.__value__)))
    ) {
      onOmitField();
    }

    onBlurProp?.(event);
  };

  return (
    <FieldTemplate
      name={name}
      label={fieldLabel}
      description={fieldDescription}
      annotations={annotations}
      touched={touched}
      className={cx({ "mb-0": hideLabel })} // Remove bottom margin if we're already hiding the label
      as={widgetsRegistry.TemplateToggleWidget}
      inputModeOptions={inputModeOptions}
      setFieldDescription={updateFieldDescription}
      onBlur={onBlur}
      {...restProps}
      // Pass in schema after spreading props to override the non-normalized schema in props
      schema={normalizedSchema}
    />
  );
};

export default BasicSchemaField;
