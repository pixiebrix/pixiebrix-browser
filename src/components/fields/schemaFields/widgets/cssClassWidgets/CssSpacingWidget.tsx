/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo, useState } from "react";
import { type SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { type InputModeOption } from "@/components/fields/schemaFields/widgets/templateToggleWidgetTypes";
import styles from "./CssSpacingWidget.module.scss";
import { UnstyledButton } from "@/components/UnstyledButton";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCaretDown, faCaretRight } from "@fortawesome/free-solid-svg-icons";
import {
  calculateNextSpacing,
  extractSpacing,
  parseValue,
} from "@/components/fields/schemaFields/widgets/cssClassWidgets/utils";
import {
  type Spacing,
  type Value,
} from "@/components/fields/schemaFields/widgets/cssClassWidgets/types";
import { useField } from "formik";

import Select from "react-select";
import { type Option } from "@/components/form/widgets/SelectWidget";
import cx from "classnames";

export interface CssSpacingWidgetControls {
  margin: boolean;
  padding: boolean;
}

const defaultOptions: CssSpacingWidgetControls = {
  margin: true,
  padding: true,
};

const spacingSides = [
  { label: "Top", side: "t" },
  { label: "Right", side: "r" },
  { label: "Bottom", side: "b" },
  { label: "Left", side: "l" },
];

const paddingOptions: Option[] = [
  { label: "0", value: "0" },
  { label: "1", value: "1" },
  { label: "2", value: "2" },
  { label: "3", value: "3" },
  { label: "4", value: "4" },
  { label: "5", value: "5" },
];

const marginOptions: Option[] = [
  { label: "0", value: "0" },
  { label: "1", value: "1" },
  { label: "2", value: "2" },
  { label: "3", value: "3" },
  { label: "4", value: "4" },
  { label: "5", value: "5" },
  { label: "-5", value: "-5" },
  { label: "-4", value: "-4" },
  { label: "-3", value: "-3" },
  { label: "-2", value: "-2" },
  { label: "-1", value: "-1" },
];

function getValue(options: Option[], spacing: Spacing[], side: string | null) {
  const numberValue = spacing.find((x) => x.side === side)?.size;
  return (
    options.find((option) => option.value === numberValue?.toString()) ?? null
  );
}

const SpacingControl: React.VFC<{
  prefix: string;
  label: string;
  className?: string;
  classes: string[];
  disabled: boolean;
  onUpdate: (update: Spacing) => void;
  options: Option[];
  placeholder: string;
}> = ({
  prefix,
  label,
  className,
  classes,
  disabled,
  onUpdate,
  options,
  placeholder,
}) => {
  const [expand, setExpand] = useState(false);

  const spacing = useMemo(
    () => extractSpacing(prefix, classes),
    [prefix, classes]
  );

  return (
    <div className={className}>
      <div className={styles.spacingControlContainer}>
        <UnstyledButton
          className={styles.expandButton}
          onClick={() => {
            setExpand(!expand);
          }}
        >
          <label htmlFor={`${prefix}-input`}>
            {label}&nbsp;
            <FontAwesomeIcon icon={expand ? faCaretDown : faCaretRight} />
          </label>
        </UnstyledButton>
        <div className="ml-5">
          <Select
            inputId={`${prefix}-input`}
            options={options}
            value={getValue(options, spacing, null)}
            className={styles.spacingControlSelect}
            isClearable
            isDisabled={disabled}
            menuPlacement="auto"
            placeholder={placeholder}
            onChange={(event) => {
              onUpdate({
                side: null,
                size: event ? Number(event.value) : null,
              });
            }}
          />
        </div>
      </div>
      {expand && (
        <div className="pl-5">
          {spacingSides.map((direction) => (
            <div
              key={direction.side}
              className={styles.spacingControlContainer}
            >
              <label htmlFor={`${prefix}-${direction.side}-input`}>
                {direction.label}
              </label>
              <Select
                options={options}
                inputId={`${prefix}-${direction.side}-input`}
                value={getValue(options, spacing, direction.side)}
                className={styles.spacingControlSelect}
                isClearable
                isDisabled={disabled}
                menuPlacement="auto"
                placeholder={placeholder}
                onChange={(option) => {
                  onUpdate({
                    side: direction.side,
                    size: option ? Number(option.value) : null,
                  });
                }}
              />
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

const CssSpacingWidget: React.VFC<
  SchemaFieldProps & { inputModeOptions: InputModeOption[] }
> = (props) => {
  const { name, uiSchema } = props;

  const controlOptions: CssSpacingWidgetControls = {
    ...defaultOptions,
    ...uiSchema?.["ui:options"],
  };

  const [{ value }, , { setValue }] = useField<Value>(name);

  const { classes, isVar, includesTemplate } = useMemo(
    () => parseValue(value),
    [value]
  );

  const disableControls = isVar || includesTemplate;

  return (
    (controlOptions.margin || controlOptions.padding) && (
      <div className={cx("d-flex", styles.spacingControlWrapper)}>
        {controlOptions.margin && (
          <SpacingControl
            prefix="m"
            label="Margin"
            placeholder="Select -5 to 5"
            classes={classes}
            disabled={disableControls}
            options={marginOptions}
            onUpdate={async (option) => {
              try {
                await setValue(calculateNextSpacing(value, "m", option));
              } catch (error) {
                reportError(error);
              }
            }}
          />
        )}

        {controlOptions.padding && (
          <SpacingControl
            prefix="p"
            label="Padding"
            placeholder="Select 0 to 5"
            classes={classes}
            disabled={disableControls}
            options={paddingOptions}
            onUpdate={async (option) => {
              try {
                await setValue(calculateNextSpacing(value, "p", option));
              } catch (error) {
                reportError(error);
              }
            }}
          />
        )}
      </div>
    )
  );
};

export default CssSpacingWidget;
