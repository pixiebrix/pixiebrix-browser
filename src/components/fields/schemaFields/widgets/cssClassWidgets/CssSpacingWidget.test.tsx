/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { type Expression } from "@/types/runtimeTypes";
import { render, screen } from "@/pageEditor/testHelpers";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import CssSpacingWidget from "@/components/fields/schemaFields/widgets/cssClassWidgets/CssSpacingWidget";
import selectEvent from "react-select-event";
import userEvent from "@testing-library/user-event";
import { getCssClassInputFieldOptions } from "@/components/fields/schemaFields/CssClassField";

const renderWidget = (value: string | Expression) =>
  render(
    <CssSpacingWidget
      inputModeOptions={getCssClassInputFieldOptions()}
      schema={{
        type: "string",
      }}
      label="spacing"
      name="cssClass"
    />,
    { initialValues: { cssClass: value } }
  );

beforeAll(() => {
  registerDefaultWidgets();
});

describe("CssClassWidget", () => {
  it("should render the spacing widget", () => {
    expect(renderWidget).toMatchSnapshot();
  });

  it("expand button works", async () => {
    renderWidget("");

    expect(
      screen.queryByRole("combobox", { name: "Top" })
    ).not.toBeInTheDocument();

    await userEvent.click(screen.getByRole("button", { name: "Margin" }));
    expect(screen.getByRole("combobox", { name: "Top" })).toBeInTheDocument();
  });

  it("clears the direction inputs when the main input is set", async () => {
    const { getFormState } = renderWidget("test mr-1 mb-1");
    await selectEvent.select(
      screen.getByRole("combobox", { name: "Margin" }),
      "1"
    );

    expect(getFormState()).toStrictEqual({ cssClass: "test m-1" });
  });

  it("clears the main input when the direction input is set", async () => {
    const { getFormState } = renderWidget("test m-1");

    await userEvent.click(screen.getByRole("button", { name: "Margin" }));

    await selectEvent.select(
      screen.getByRole("combobox", { name: "Top" }),
      "3"
    );

    expect(getFormState()).toStrictEqual({ cssClass: "test mt-3" });
  });
});
