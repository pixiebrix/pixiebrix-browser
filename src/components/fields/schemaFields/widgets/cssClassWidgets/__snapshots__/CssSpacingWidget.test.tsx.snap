// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CssClassWidget should render the spacing widget 1`] = `[Function]`;
